src/components/fields/NumberField.js

```diff
@@ -3,14 +3,71 @@ import React from "react";
 import * as types from "../../types";
 import { asNumber } from "../../utils";
 
-function NumberField(props) {
-  const { StringField } = props.registry.fields;
-  return (
-    <StringField
-      {...props}
-      onChange={value => props.onChange(asNumber(value))}
-    />
-  );
+/**
+ * The NumberField class has some special handling for dealing with trailing
+ * decimal points and/or zeroes. This logic is designed to allow trailing values
+ * to be visible in the input element, but not be represented in the
+ * corresponding form data.
+ *
+ * The algorithm is as follows:
+ *
+ * 1. When the input value changes the value is cached in the component state
+ *
+ * 2. The value is then normalized, removing trailing decimal points and zeros,
+ *    then passed to the "onChange" callback
+ *
+ * 3. When the component is rendered, the formData value is checked against the
+ *    value cached in the state. If it matches the cached value, the cached
+ *    value is passed to the input instead of the formData value
+ */
+class NumberField extends React.Component {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      lastValue: props.value,
+    };
+  }
+
+  handleChange = value => {
+    // Cache the original value in component state
+    this.setState({ lastValue: value });
+
+    // Check that the value is a string (if the widget used is a select (due to
```

Comment:
@LucianBuzzo can you fix the parentheses here?

Also, what is the scenario, exactly, in which there's a select with type=number with an option that has a trailing decimal point or multiple zeroes? Won't the schema never have these? Maybe if you can add a test for this it would be clearer.

Reply:
@epicfaace This logic needs to be here, because as you type in a value such as `0.001` each keystroke will trigger a change event, so the input will have to handle inputs like `0.00` as the user is typing. Additionally, you may have someone copy and paste values like `201.00` into the form, which is a valid number that you would expect to be processed correctly.